Readme.md

# response-send

  response.send() prototype method

## License 

(The MIT License)

Copyright (c) 2012 TJ Holowaychuk &lt;tj@vision-media.ca&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.1.0 / 2012-07-04 
==================

  * add length to etags

0.0.1 / 2010-01-03
==================

  * Initial release


test/send.js

var http = require('http')
  , send = require('..')
  , server = http.createServer
  , request = require('supertest')
  , pending = require('./utils/pending');

// #nodejsWTF?

http.ServerResponse.prototype.__defineGetter__('req', function(){
  return this.socket.parser.incoming;
});

// augment prototype

http.ServerResponse.prototype.send = send;
http.ServerResponse.prototype.json = send.json({
  spaces: 2
});

// shared

function shared(fn) {
  it('should not override a previous Content-Type', function(done){
    var app = server(function(req, res){
      res.setHeader('Content-Type', 'image/png');
      fn(req, res);
    });

    request(app)
    .get('/')
    .expect('Content-Type', 'image/png')
    .end(done);
  })
}

// tests

describe('res.send()', function(){
  it('should support HEAD', function(done){
    var app = server(function(req, res){
      res.send('Hello');
    });

    request(app)
    .head('/')
    .expect('')
    .expect('Content-Length', '5')
    .expect(200, done);
  })

  it('should add a weak ETag', function(done){
    done = pending(3, done);

    var app = server(function(req, res){
      switch (req.url) {
        case '/foo':
          return res.send(Array(1000).join('foo'));
        case '/bar':
          return res.send(Array(1000).join('bar'));
      }
    });

    request(app)
    .get('/foo')
    .expect('ETag', '2997.601152967')
    .end(done);

    request(app)
    .get('/foo')
    .expect('ETag', '2997.601152967')
    .end(done);

    request(app)
    .get('/bar')
    .expect('ETag', '2997.-1124164645')
    .end(done);
  })

  it('should respond with 304 when ETag matches', function(done){
    done = pending(2, done);

    var app = server(function(req, res){
      res.send(Array(1000).join('foo'));
    });

    request(app)
    .get('/foo')
    .expect('ETag', '2997.601152967')
    .expect(200, done);

    request(app)
    .get('/foo')
    .set('If-None-Match', '2997.601152967')
    .end(function(err, res){
      res.should.have.status(304);
      res.headers.should.not.have.property('content-length');
      res.headers.should.not.have.property('content-type');
      res.text.should.equal('');
      done();
    });
  })

  it('should pass through non-2xx or 304 responses', function(done){
    done = pending(2, done);

    var app = server(function(req, res){
      res.statusCode = 500;
      res.send(Array(1000).join('foo'));
    });

    request(app)
    .get('/foo')
    .expect('ETag', '2997.601152967')
    .expect(500, done);

    request(app)
    .get('/foo')
    .set('If-None-Match', '2997.601152967')
    .end(function(err, res){
      res.should.have.status(500);
      res.headers.should.have.property('content-length');
      res.headers.should.have.property('content-type');
      res.text.should.equal(Array(1000).join('foo'));
      done();
    });
  })
})

describe('res.send(null)', function(){
  it('should respond with ""', function(done){
    var app = server(function(req, res){
      res.send(null);
    });

    request(app)
    .get('/')
    .expect('', done);
  })
})

describe('res.send(status)', function(){
  it('should respond the status code string', function(done){
    var app = server(function(req, res){
      res.send(201);
    });

    request(app)
    .get('/')
    .expect(201, 'Created', done);
  })

  shared(function(req, res){
    res.send(201);
  });
})

describe('res.send(Buffer)', function(){
  it('should respond with an octet-stream', function(done){
    var app = server(function(req, res){
      res.send(new Buffer('hey'));
    });

    request(app)
    .get('/')
    .expect('hey')
    .expect('Content-Length', '3')
    .expect('Content-Type', 'application/octet-stream')
    .expect(200, done);
  })

  shared(function(req, res){
    res.send(new Buffer('hey'));
  });
})

describe('res.send(string)', function(){
  it('should respond with html', function(done){
    var app = server(function(req, res){
      res.send('<p>Hello</p>');
    });

    request(app)
    .get('/')
    .expect('<p>Hello</p>')
    .expect('Content-Length', '12')
    .expect('Content-Type', 'text/html')
    .expect(200, done);
  })

  it('should respond with byte-length', function(done){
    var app = server(function(req, res){
      res.send('…');
    });

    request(app)
    .get('/')
    .expect('…')
    .expect('Content-Length', '3')
    .expect(200, done);
  })

  shared(function(req, res){
    res.send('<p>Hello</p>');
  });
})

describe('res.send(status, string)', function(){
  it('should respond with html', function(done){
    var app = server(function(req, res){
      res.send(201, '<p>Created</p>');
    });

    request(app)
    .get('/')
    .expect('<p>Created</p>')
    .expect('Content-Type', 'text/html')
    .expect(201, done);
  })

  shared(function(req, res){
    res.send(201, '<p>Created</p>');
  });
})

describe('res.send(string, status)', function(){
  it('should respond with html', function(done){
    var app = server(function(req, res){
      res.send('<p>Created</p>', 201);
    });

    request(app)
    .get('/')
    .expect('<p>Created</p>')
    .expect('Content-Type', 'text/html')
    .expect(201, done);
  })

  shared(function(req, res){
    res.send('<p>Created</p>', 201);
  });
})

describe('res.send(String)', function(){
  it('should respond with html', function(done){
    var app = server(function(req, res){
      res.send(new String('<p>Hello</p>'));
    });

    request(app)
    .get('/')
    .expect('<p>Hello</p>')
    .expect('Content-Length', '12')
    .expect('Content-Type', 'text/html')
    .expect(200, done);
  })
})

describe('res.send(Object)', function(){
  it('should respond with json', function(done){
    var app = server(function(req, res){
      res.send({ name: 'tobi' });
    });

    request(app)
    .get('/')
    .expect('{\n  "name": "tobi"\n}')
    .expect('Content-Length', '20')
    .expect('Content-Type', 'application/json')
    .expect(200, done);
  })
})

describe('res.send(status, Object)', function(){
  it('should respond with json', function(done){
    var app = server(function(req, res){
      res.send(201, { name: 'tobi' });
    });

    request(app)
    .get('/')
    .expect('{\n  "name": "tobi"\n}')
    .expect('Content-Length', '20')
    .expect('Content-Type', 'application/json')
    .expect(201, done);
  })
})

describe('res.send(Object, status)', function(){
  it('should respond with json', function(done){
    var app = server(function(req, res){
      res.send({ name: 'tobi' }, 201);
    });

    request(app)
    .get('/')
    .expect('{\n  "name": "tobi"\n}')
    .expect('Content-Length', '20')
    .expect('Content-Type', 'application/json')
    .expect(201, done);
  })
})

test/utils/pending.js

module.exports = function(n, fn){
  return function(err){
    if (err) return fn(err);
    --n || fn();
  }
}

lib/response-send.js

/**
 * Module dependencies.
 */

var crc = require('crc')
  , http = require('http')
  , fresh = require('fresh');

/**
 * Send a response.
 *
 * Examples:
 *
 *     res.send(new Buffer('wahoo'));
 *     res.send({ some: 'json' });
 *     res.send('<p>some html</p>');
 *     res.send(404, 'Sorry, cant find that');
 *     res.send(404);
 *
 * @param {Mixed} body or status
 * @param {Mixed} body
 * @return {ServerResponse}
 * @api public
 */

exports = module.exports = function(body){
  var req = this.req
    , method = req.method
    , len;

  // allow status / body
  if (2 == arguments.length) {
    // res.send(body, status) backwards compat
    if ('number' != typeof body && 'number' == typeof arguments[1]) {
      this.statusCode = arguments[1];
    } else {
      this.statusCode = body;
      body = arguments[1];
    }
  }

  // null
  if (null == body) body = '';

  // convert primitives
  body = body.valueOf();

  switch (typeof body) {
    // response status
    case 'number':
      if (!this.getHeader('Content-Type')) this.setHeader('Content-Type', 'text/plain');
      this.statusCode = body;
      body = http.STATUS_CODES[body];
      break;
    // string defaulting to html
    case 'string':
      if (!this.getHeader('Content-Type')) {
        this.charset = this.charset || 'utf-8';
        this.setHeader('Content-Type', 'text/html');
      }
      break;
    case 'boolean':
    case 'object':
      if (null == body) {
        body = '';
      } else if (Buffer.isBuffer(body)) {
        if (!this.getHeader('Content-Type')) this.setHeader('Content-Type', 'application/octet-stream');
      } else {
        return this.json(body);
      }
      break;
  }

  // populate Content-Length
  if (undefined !== body && !this.getHeader('Content-Length')) {
    this.setHeader('Content-Length', len = Buffer.isBuffer(body)
      ? body.length
      : Buffer.byteLength(body));
  }

  // ETag support
  // TODO: W/ support
  if (len > 1024) {
    var etag = Buffer.isBuffer(body) ? crc.buffer.crc32(body) : crc.crc32(body);
    if (!this.getHeader('ETag')) this.setHeader('ETag', len + '.' + etag);
  }

  // determine if it's cacheable
  var cache = 'GET' == method || 'HEAD' == method;
  if (cache) cache = (this.statusCode >= 200 && this.statusCode < 300) || 304 == this.statusCode;

  // freshness
  if (fresh(req.headers, this._headers) && cache) {
    this.statusCode = 304;
  }

  // strip irrelevant headers
  if (204 == this.statusCode || 304 == this.statusCode) {
    this.removeHeader('Content-Type');
    this.removeHeader('Content-Length');
    body = '';
  }

  // respond
  this.end('HEAD' == method ? null : body);
  return this;
};

/**
 * Send JSON response.
 *
 * Examples:
 *
 *     res.json(null);
 *     res.json({ user: 'tj' });
 *     res.json(500, 'oh noes!');
 *     res.json(404, 'I dont have that');
 *
 * @param {Mixed} obj or status
 * @param {Mixed} obj
 * @return {ServerResponse}
 * @api public
 */

exports.json = function(options){
  options = options || {};
  return function(obj){
    // allow status / body
    if (2 == arguments.length) {
      // res.json(body, status) backwards compat
      if ('number' == typeof arguments[1]) {
        this.statusCode = arguments[1];
      } else {
        this.statusCode = obj;
        obj = arguments[1];
      }
    }

    // settings
    var app = this.app
      , replacer = options.replacer
      , spaces = options.spaces
      , body = JSON.stringify(obj, replacer, spaces);

    // content-type
    this.charset = this.charset || 'utf-8';
    this.setHeader('Content-Type', 'application/json');

    return this.send(body);
  }
};